### PR TITLE
Gavin/212 add pgsync to keep elasticsearch up to date with postgres

### LINKED ISSUE #212: Add PGSync to keep elasticsearch up to date with Postgres
## Feature Description
I'm going to try and use [PGSync](https://pgsync.com/) to keep elasticsearch up to date with Postgres (the source of truth) automatically.

relates: #70 

## Problem You're Trying to Solve
After a discussion with Max, I realized that making backend keep Elasticsearch up to date by hitting the endpoint whenever a listing gets added is risky if either system goes down. 

Additionally, the 2000 or so demo listings inserted into Postgres before the demo are not being sent to Elasticsearch as is.

### PR DESCRIPTION
**Pull request by GDeane**
# Description
I want to use [PGSync](https://pgsync.com/)  to keep elasticsearch in sync with Postgres (reading from postgres and writing to elasticsearch).

I was working in the martletplace repository on the search endpoint and I have realized a problem after chatting with Max.

It is very easy for the elasticsearch index which search depends upon to get out of sync with the Postgres database (the source of truth).

Originally, I had planned to forward Listings updates from the backend to the search endpoint using the /api/search/reindex/listing-created endpoint (and two others, which are present in the search endpoint)

Unfortunately, this has issues with staying in sync particularly when either service can potentially go down.

Closes #212

## How to Test
1. Run `docker-compose up -- build`

2. In a browser, paste `http://localhost/api/search?query=chocolate&minPrice=5&maxPrice=10&status=AVAILABLE&searchType=LISTINGS&latitude=48.4633&longitude=-123.3108&sort=RELEVANCE&page=1&limit=6`
expected output:
image

3. In a browser, paste `http://localhost/api/search?query=user&status=SOLD&searchType=USERS&latitude=48.4633&longitude=-123.3108&sort=RELEVANCE&page=1&limit=6`
expected output:
image

## Checklist
- [X] The code includes tests if relevant
- [X] I have *actually* self-reviewed my changes and done QA

### CODE REVIEW

File: apps/data/pgsync/pgsync.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Configuration for PGSync
+database_url: postgres://martletplace:johuigytfrdcv@datalayer:8300/martletplace
+elasticsearch_url: https://elastic:serxdfcghjfc@elasticsearch:8311
```
Comment (MNThomson) [inline apps/data/pgsync/pgsync.yml:3]: Hoping there's a method to use env vars

File: docker-compose.yml [hunk 2]
```diff
@@ -19,7 +19,7 @@ services:
     volumes: [ ./apps/proxy:/app ]
     environment: []
     ports: [ 80:80, 443:443 ]
-    networks: 
+    networks:
```
Comment (MNThomson) [inline docker-compose.yml:22]: WHAT DID YOU DO??? HOW???

File: docker-compose.yml [hunk 3]
```diff
@@ -157,6 +157,23 @@ services:
...
     mem_limit: 1GB
 
+  pgsync:
+    build: apps/data/pgsync
+    container_name: martletplace_pgsync
+    depends_on:
+      - database
+      - elasticsearch
+    volumes:
+      - ./apps/data/pgsync/config/pgsync.yml:/app/config/pgsync.yml
```
Comment (MNThomson) [inline docker-compose.yml:167]: Not sure there's any point in mounting this file in unless pgsync hot reloads

File: docker-compose.yml [hunk 4]
```diff
@@ -157,6 +157,23 @@ services:
...
+    environment:
+      - DB_ENDPOINT=postgres://martletplace:johuigytfrdcv@datalayer:8300/martletplace
+      - ES_ENDPOINT=https://elastic:serxdfcghjfc@elasticsearch:8311
+      - PG_USER=martletplace
+      - PG_PASSWORD=johuigytfrdcv
```
Comment (MNThomson) [inline docker-compose.yml:172]:
```suggestion
```

File: docker-compose.yml [hunk 5]
```diff
@@ -157,6 +157,23 @@ services:
...
+    networks:
+      - datalayer-database
+      - datalayer-elasticsearch
```
Comment (MNThomson) [inline docker-compose.yml:175]: This should be connected to

```suggestion
      - pgsync-database
      - pgsync-elasticsearch
```

And then you'll need to add one of those to the two databases respectively

File: apps/data/pgsync/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,3 @@
+pgsync
+
+## Dev Dependencies
```
Comment (MNThomson) [inline apps/data/pgsync/requirements.txt:3]:
```suggestion
```

**Comment by GDeane** [comment]
I'm running into the problem where elasticsearch requires location fields to be mapped as geo_point with the fields "lat" and "lon" in order to search by location (https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-point.html). 

previously, location fields looked like this:
`"location": {
            "lat": 45.4215,
            "lon": -75.6972
          },`

Now, they look like this after using PGSync:
`
"location": {
            "latitude": 40.7128,
            "longitude": -74.006
          },
`

When I search by location, I currently get this error:
image.

In short, if location can't be changed to "lat" and "lon" instead of "latitude" and "longitude" using the PGSync schema (or some other method), we won't be able to search within 5 km of or sort by location.


Meanwhile, PGSync has the issue where it does not recognize the LOCATION_TYPE defined in PGSync
image

I have not been able to figure out how to resolve this.

File: docker-compose.yml [hunk 7]
```diff
@@ -146,9 +147,38 @@ services:
     volumes: []
     environment:
       - ELASTIC_PASSWORD=serxdfcghjfc
-    networks: [ datalayer-elasticsearch, search-datalayer ]
+    networks: [ datalayer-elasticsearch, pgsync-elasticsearch, search-elasticsearch ]
```
Comment (MNThomson) [inline docker-compose.yml:150]: This is clearly the wrong networks, but if everything works we can fix this later

File: docker-compose.yml [hunk 8]
```diff
@@ -146,9 +147,38 @@ services:
...
+      - PG_PASSWORD=johuigytfrdcv
+      - ELASTICSEARCH_HOST=elasticsearch
+      - ELASTICSEARCH_PORT=8311
+      - ELASTICSEARCH_USER=elastic
+      - ELASTICSEARCH_PASSWORD=serxdfcghjfc
+      - ELASTICSEARCH_SCHEME=https
+      - ELASTICSEARCH_VERIFY_CERTS=False
+      - REDIS_HOST=cache
+      - REDIS_PORT=8313
+      - REDIS_DB=0
```
Comment (MNThomson) [inline docker-compose.yml:176]: These need to be cleaned up later

File: apps/data/pgsync/schema.json [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+                            "parent": [
+                                "seller_id"
+                            ]
+                        }
+                    }
+                }
+            ],
+            "transform":
+            {
+                "rename":
```
Comment (MNThomson) [inline apps/data/pgsync/schema.json:40]: I'm assuming we can use another rename to change from latitude -> lat, etc. Sounds like a problem for later

**Review by MNThomson — APPROVED** [review]

**Review by emilasuska — APPROVED** [review]
LGTM

File: docker-compose.yml (hunk 7 above)
Comment (GDeane) [reply docker-compose.yml:150]: gotcha

File: docker-compose.yml (hunk 8 above)
Comment (GDeane) [reply docker-compose.yml:176]: amen brotha

File: apps/data/pgsync/schema.json (hunk 9 above)
Comment (GDeane) [reply apps/data/pgsync/schema.json:40]: I tried everything I could find in the PGSync docs to do this. I think it PGSync doesn't properly support Postgres type definitions so it can't do the rename as part of the sync operation.
